WIP: Add library element context menu

dbrgn opened the pull request:
Add context menu entry for removing the currently selected element.

Refs #200.

@ubruhin this is a first draft. Is the way I approached the issue correct? If yes, I can add support for the other library elements.

This could also be extended later on, e.g. when right-clicking on the empty space of the symbol list widget, we could add an entry called "Create symbol". When right-clicking on an existing symbol, we could add an entry called "Create symbol from this symbol" or something like that.

I think once this is in place, we can actually remove the delete menu entries, I think they're bad UX because it's not clear what they're referring to.

ubruhin commented on libs/librepcb/libraryeditor/lib/libraryoverviewwidget.cpp:
```diff
@@ -103,6 +103,14 @@ LibraryOverviewWidget::LibraryOverviewWidget(const Context&          context,
           this, &LibraryOverviewWidget::commitMetadata);
 
+  // Set up context menu triggers
+  mUi->lstSym->setContextMenuPolicy(Qt::CustomContextMenu);
```
This property can be set in the designer.

ubruhin commented on libs/librepcb/libraryeditor/lib/libraryoverviewwidget.cpp:
```diff
@@ -135,6 +143,20 @@ bool LibraryOverviewWidget::save() noexcept {
+void LibraryOverviewWidget::symShowContextMenu(const QPoint &pos) noexcept {
+    QPoint globalPos = mUi->lstSym->mapToGlobal(pos);
+    QMenu menu;
+    menu.addAction(QIcon(":/img/actions/delete.png"), tr("Remove"), this, &LibraryOverviewWidget::symRemove);
```
I wouldn't use a separate function and signal/slot connection for `QAction` handlers. I think it's simpler to get the return value of `QMenu::exec()` to decide what action was triggered. Then you can also use [`QListWidget::itemAt()`](http://doc.qt.io/qt-5/qlistwidget.html#itemAt) to get the item which was actually right-clicked. Later we need it anyway to decide which context menu entries should be shown and which not...

See for example `SymbolEditorState_Select::openContextMenuAtPos()`.

ubruhin commented:
> @ubruhin this is a first draft. Is the way I approached the issue correct? If yes, I can add support for the other library elements.

Looks not bad so far :+1: Just don't forget that a library element may be opened at the time of its deletion, in that case the corresponding tab needs to be closed first.

> This could also be extended later on, e.g. when right-clicking on the empty space of the symbol list widget, we could add an entry called "Create symbol". When right-clicking on an existing symbol, we could add an entry called "Create symbol from this symbol" or something like that.

:+1:

> I think once this is in place, we can actually remove the delete menu entries, I think they're bad UX because it's not clear what they're referring to.

Would be nice if you could do it in this PR :)

dbrgn commented:
@ubruhin do you know how to prevent the following bug?

https://tmp.dbrgn.ch/librepcb-2018-11-18_23.06.34.webm

For some reason, the selection is not updated sometimes when right-clicking on another item while the context menu for the first item is still open.

dbrgn commented:
Regarding closing the tab if it's open:

- This is something I'd have to do in the libraryeditor.cpp, right? So I'd
  register a slot there, and emit an event after deleting the item in
  libraryoverviewwidget.cpp?
- How do I find out whether the element was opened in a tab?

I removed the remove action in the file menu, it's now replaced with the context menu. However, regarding the disabling of the delete/rotate items in the edit menu and toolbar, I think that's complex enough that it should be done in a separate PR. It doesn't have anything to do with removing library elements.

dbrgn commented:
PR is now refactored and uses `itemAt` to get the currently selected item. Furthermore, instead of the sender,  he selected item text as well as the file path are passed to the `removeSelectedItem` method.

ubruhin commented on libs/librepcb/libraryeditor/lib/libraryoverviewwidget.cpp:
```diff
@@ -254,6 +268,64 @@ void LibraryOverviewWidget::updateElementList(QListWidget& listWidget,
+  QListWidgetItem* selectedItem = list->itemAt(pos);
+
+  // Get item text
+  const QString itemName = selectedItem->text();
```
This line causes a crash if no item is selected.

ubruhin commented on libs/librepcb/libraryeditor/lib/libraryoverviewwidget.cpp:
```diff
@@ -254,6 +268,64 @@ void LibraryOverviewWidget::updateElementList(QListWidget& listWidget,
+      QMessageBox::Yes, QMessageBox::Cancel);
+  if (ret == QMessageBox::Yes) {
+    try {
+      FileUtils::removeDirRecursively(itemPath);
```
If the selected element is currently opened in the library editor (as a tab), it must be closed before removing its directory.

ubruhin commented on libs/librepcb/libraryeditor/libraryeditor.h:
```diff
@@ -139,6 +138,7 @@ class LibraryEditor final : public QMainWindow,
   void zoomOutTriggered() noexcept;
   void zoomAllTriggered() noexcept;
   void editGridPropertiesTriggered() noexcept;
+
```
Newline added by accident?

ubruhin commented on libs/librepcb/libraryeditor/libraryeditor.h:
```diff
@@ -148,6 +148,11 @@ class LibraryEditor final : public QMainWindow,
+
+  void removeSymbolTriggered(const FilePath& fp) noexcept;
+  void removeComponentTriggered(const FilePath& fp) noexcept;
+  void removeLibraryElementTriggered(const QString& elementType, const FilePath& fp) noexcept;
```
These methods do not exist in the *.cpp.

dbrgn commented:
Ok, only issue left is now the closing of the open tab.

dbrgn commented:
Ok, all requested issues fixed. If you want you can already review, but I also want to add a functional test :slightly_smiling_face:

ubruhin commented on libs/librepcb/libraryeditor/libraryeditor.h:
```diff
@@ -126,10 +126,12 @@ class LibraryEditor final : public QMainWindow,
   LibraryEditor& operator=(const LibraryEditor& rhs) = delete;
 
+public slots:
+  void closeTabIfOpen(const FilePath& fp) noexcept;
```
This method does not need to be public, it's never called (directly) from outside this class. In addition, it does not need to be marked as a slot, since C++11 also "normal" methods can be used as a slot when using the overload of `QObject::connect()` which takes function pointers. So I would suggest to move this method below `editDeviceTriggered()`, then the order corresponds to the `connect()` calls in the constructor :)

ubruhin commented:
Ah, I forgot: If you have clang-format 6.x installed, it would be nice if you could run `./dev/format_code.sh` to format the code ;)

dbrgn commented:
Ok, all requested issues fixed. If you want you can already review, but I also want to add a functional test :slightly_smiling_face:

dbrgn commented:
Functional tests are waiting for https://github.com/parkouss/funq/pull/47, so I think this can be merged.
